piptools/utils.py

```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
             markers=parts.markers,
             extras=parts.extras,
         )
+
+
+def callback_config_file_defaults(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> str | None:
+    """
+    Returns the path to the config file with defaults being used, or `None` if no such file is
+    found.
+
+    Defaults for `click.Command` parameters should be override-able in a config file. `pip-tools`
+    will use the first file found, searching in this order: an explicitly given config file, a
+    `.pip-tools.toml`, a `pyproject.toml` file. Those files are searched for in the same directory
+    as the requirements input file.
+    """
+    if value is None:
+        config_file = select_config_file(ctx.params.get("src_files", ()))
+        if config_file is None:
+            return None
+    else:
+        config_file = Path(value)
+
+    try:
+        config = parse_config_file(config_file)
+    except OSError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not read '{config_file}': {e}"
+        )
+    except ValueError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not parse '{config_file}': {e}"
+        )
+
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
```

Comment:
Why is this needed? Would the following be equivalent?
```suggestion
    if ctx.default_map is None:
        ctx.default_map = {}
    ctx.default_map.update(config)
```

FWIW, pieces like this could benefit from being in a dedicated function with a descriptive name.

Reply:
I didn't have an explicit reason for it - I was following a suggestion in the PR to look at the way black implemented this as a click callback and used their code as an inspiration, in which they reconstruct a whole new dict for `default_map` instead of updating the current one. The `git blame` seems to indicate the suggested change you've offered would run afoul of the type-checker.

https://github.com/psf/black/blame/b2ee211b5ad84b62738ac0997b73bf6ee9a74d06/src/black/__init__.py#L129-L134C1

Reply:
Honestly, I wouldn't recommend doing whatever black does, as it has countless shortcomings. To me, it's rather an indication that an idea might be bad.
I understand that in-browser code suggestions often need to be corrected later on, locally. This is still a good way of visualizing the expectations/ideas/suggestions...

Reply:
OTOH, I was making the suggestions based on the engineering thinking rather than whether some other project does something in one way or another. Keeping a project easy-to-read/follow and maintainable sometimes requires more attention to detail. As I was progressing with the review, my perception, and understanding changed/became more accurate, which is why I added new suggestions on top of the previous ones. Making pauses helped me gain better perspective..

Reply:
I'll try to see if that 3-year old problem got solved by now..

Reply:
> The `git blame` seems to indicate the suggested change you've offered would run afoul of the type-checker.

Good news — mypy didn't explode on this.